pkg/hub/controllers/cluster/reconciler_test.go

```diff
@@ -0,0 +1,284 @@
+		mock.IsType(&rbacv1.ClusterRole{}),
+		mock.IsType([]k8sclient.CreateOption(nil)),
+	).Return(nil)
+	client.On("Create",
+		mock.IsType(ctx),
+		mock.IsType(&rbacv1.ClusterRoleBinding{}),
+		mock.IsType([]k8sclient.CreateOption(nil)),
+	).Return(nil)
+	client.On("Create",
+		mock.IsType(ctx),
+		mock.IsType(&corev1.ConfigMap{}),
+		mock.IsType([]k8sclient.CreateOption(nil)),
+	).Return(nil)
+	client.On("Get",
+		mock.IsType(ctx),
+		mock.IsType(types.NamespacedName{Name: deregisterJobName, Namespace: ControlPlaneNamespace}),
+		mock.IsType(&batchv1.Job{}),
+	).Return(nil)
+	client.On("Delete",
+		mock.IsType(ctx),
+		mock.IsType(&batchv1.Job{}),
+		mock.IsType([]k8sclient.DeleteOption{}),
+	).Return(nil)
+	client.On("Create",
+		mock.IsType(ctx),
+		mock.IsType(&batchv1.Job{}),
+		mock.IsType([]k8sclient.CreateOption(nil)),
+	).Return(nil)
+
+	_, _, err := reconciler.handleClusterDeletion(testClusterObjWithFinalizer, ctx, expected.req)
+	if expected.err != err {
+		t.Error("Expected error:", expected.err, " but got ", err)
+	}
+}
+
+func TestReconcilerToFailWhileCallingCreateDeregisterJob(t *testing.T) {
+	expected := struct {
+		ctx    context.Context
+		req    reconcile.Request
+		res    reconcile.Result
+		errMsg string
+	}{
+
+		context.WithValue(context.Background(), types.NamespacedName{Namespace: "kube-slice", Name: "kube-slice"}, testClusterObjWithFinalizer),
+		reconcile.Request{NamespacedName: types.NamespacedName{Name: testClusterName, Namespace: testProjectNamespace}},
+		reconcile.Result{},
+		"error updating status of deregistration on the controller",
+	}
+	client := utilmock.NewClient()
+
+	gv := hubv1alpha1.GroupVersion
+	testScheme := runtime.NewScheme()
+	err := scheme.AddToScheme(testScheme)
+	if err != nil {
+		t.Fatalf("Error adding core scheme to test scheme: %v", err)
+	}
+	testScheme.AddKnownTypeWithName(gv.WithKind("Cluster"), &hubv1alpha1.Cluster{})
+
+	testClusterEventRecorder := mevents.NewEventRecorder(client, testScheme, ossEvents.EventsMap, mevents.EventRecorderOptions{
+		Cluster:   "test-cluster",
+		Project:   "avesha",
+		Component: "worker-operator",
+		Namespace: controllers.ControlPlaneNamespace,
+	})
+	mf, _ := metrics.NewMetricsFactory(prometheus.NewRegistry(), metrics.MetricsFactoryOptions{})
+	reconciler := NewReconciler(client, testClusterEventRecorder, mf)
+	reconciler.InjectClient(client)
+	ctx := context.WithValue(context.Background(), types.NamespacedName{Name: testClusterName, Namespace: testProjectNamespace}, testClusterObjWithFinalizer)
+	clusterKey := types.NamespacedName{Namespace: testProjectNamespace, Name: testClusterName}
+
+	client.On("Get",
+		mock.IsType(ctx),
+		mock.IsType(clusterKey),
+		mock.IsType(&hubv1alpha1.Cluster{}),
+	).Return(nil)
+	client.StatusMock.On("Update",
+		mock.IsType(ctx),
+		mock.IsType(&hubv1alpha1.Cluster{}),
+		mock.IsType([]k8sclient.UpdateOption(nil)),
+	).Return(errors.New("error updating status of deregistration on the controller"))
+	client.On("Create",
```

Comment:
Haven't we written any ITs for Cluster Deregister?
Please add the same

Reply:
We have written UT's we are getting a good coverage (90%). and as disscussed over call we can not write IT's with the current bash script approach.
We are testing the whole flow using e2e and e2e with UI too. Apart from this we have fixed all the current ITs also.

Reply:
@Rahul-D78 We can have IT atleast till JOB creation

Reply:
We tried writing the IT until job creation but that is not happening because the envtest job reconciler is not adding unique selector while creation of job which is mandatory for a job to be created. If we manually add the selector in job spec the test case is passing but manually adding the selector in job spec is not a recomended because it means we have to select a unique label on our own for the job and if we fail in doing such thing it may harm other jobs too.
It is mentioned here in the k8s docs to not mention selector spec - https://kubernetes.io/docs/concepts/workloads/controllers/job/#specifying-your-own-pod-selector

Reply:
Added a integration test case